packages/runtime/container-runtime/src/summarizerHeuristics.ts

```diff
@@ -39,46 +45,72 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
+        this.numSystemOps = 0;
+
+        this.numNonSystemOpsBefore = this.numNonSystemOps;
+        this.numNonSystemOps = 0;
     }
 
     public markLastAttemptAsSuccessful() {
         this._lastSuccessfulSummary = { ...this.lastAttempt };
     }
+
+    public adjustOnFailure(): void {
+        this.numSystemOps += this.numSystemOpsBefore;
+        this.numNonSystemOps += this.numNonSystemOpsBefore;
+    }
 }
 
 /**
  * This class contains the heuristics for when to summarize.
  */
 export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
-    private readonly idleTimer: Timer;
+    private readonly idleTimer: Timer | undefined;
+    private readonly runSummarize: (reason: SummarizeReason) => void;
 
     public constructor(
         private readonly heuristicData: ISummarizeHeuristicData,
         private readonly configuration: ISummaryConfiguration,
-        private readonly trySummarize: (reason: SummarizeReason) => void,
+        trySummarize: (reason: SummarizeReason) => void,
         private readonly minOpsForAttemptOnClose = 50,
+        private readonly summarizeStrategies: ISummarizeHeuristicStrategy[] = getDefaultSummarizeHeuristicStrategies(),
+        useIdleTimerStrategy: boolean = true,
     ) {
-        this.idleTimer = new Timer(
-            this.configuration.idleTime,
-            () => this.trySummarize("idle"));
+        if (useIdleTimerStrategy) {
+            this.idleTimer = new Timer(
+                this.configuration.idleTime,
+                () => this.runSummarize("idle"));
+        }
+
+        this.runSummarize = (reason: SummarizeReason) => {
+            this.idleTimer?.clear();
+
+            // We shouldn't attempt a summary if there are no new processed ops
+            const opsSinceLastAck = this.opsSinceLastAck;
+            if (opsSinceLastAck > 0) {
+                trySummarize(reason);
+            }
+        }
     }
 
     public get opsSinceLastAck(): number {
         return this.heuristicData.lastOpSequenceNumber - this.heuristicData.lastSuccessfulSummary.refSequenceNumber;
     }
 
+    public start() {
+        this.idleTimer?.start();
+    }
+
     public run() {
-        const timeSinceLastSummary = Date.now() - this.heuristicData.lastSuccessfulSummary.summaryTime;
-        const opsSinceLastAck = this.opsSinceLastAck;
-        if (timeSinceLastSummary > this.configuration.maxTime) {
-            this.idleTimer.clear();
-            this.trySummarize("maxTime");
-        } else if (opsSinceLastAck > this.configuration.maxOps) {
-            this.idleTimer.clear();
-            this.trySummarize("maxOps");
-        } else {
-            this.idleTimer.restart();
+        for (const strategy of this.summarizeStrategies) {
+            if (strategy.shouldRunSummarize(this.configuration, this.heuristicData)) {
+                return this.runSummarize(strategy.summarizeReason);
+            }
         }
+
+        this.idleTimer?.restart();
```

Comment:
I think this object is created on demand. That means we already have some trailing ops when we get here, and we already lost ability to separate them into user ops & system ops. We should looking into doing one of the following:
1. Assume some mix (i.e. 100% of user ops, or 70% - 30%) and initiate state appropriately
2. Figure out how to move op counting out of this class, i.e. have some part of this logic be running all the time.
   - That's essentially the same as leveraging OpTracker class. Though we would need to subtract 2 from number of system ops to ignore summarize ops - I think that's fine approach.
    -